File: flac2ogg.pl
#!/usr/bin/perl -w

use Audio::FLAC::Header;
use Audio::TagLib;

sub RemoveReplayGain
{
	foreach $ogg (@_)
	{
		$ogg->tag()->removeField(Audio::TagLib::String->new("REPLAYGAIN_ALBUM_GAIN","Latin1"));
		$ogg->tag()->removeField(Audio::TagLib::String->new("REPLAYGAIN_ALBUM_PEAK","Latin1"));
		$ogg->tag()->removeField(Audio::TagLib::String->new("REPLAYGAIN_TRACK_GAIN","Latin1"));
		$ogg->tag()->removeField(Audio::TagLib::String->new("REPLAYGAIN_TRACK_PEAK","Latin1"));
	}
}

sub MakeDirTree
{
	my $Directory=shift;
	
	my @Dirs=split(/\//,$Directory);
	
	my $Dir="";
	for $ThisDir (@Dirs)
	{
		$Dir.=$ThisDir."/";
		mkdir($Dir);
	}
}
	
sub ParseFile
{
	my $File=shift;
	my $BasePath=shift;

	$BasePath=~s/\+/\\\+/;
	$BasePath=~s/\$/\\\$/;
	$BasePath=~s/\^/\\\^/;
	$BasePath=~s/\!/\\\!/;

	if ($File =~ /${BasePath}\/(.*)\/(.*).flac/)
	{
		return ($1,$2);
	}
	else 
	{
		if ($File =~ /${BasePath}\/(.*).flac/)
		{
			return ($BasePath,$1);
		}
	}
}

sub FixName
{
	my $Name=shift;
	
	$Name=~s/\//\-/g;
	$Name=~s/\:/\-/g;
	$Name=~s/"/\-/g;
	$Name=~s/'/\-/g;
	$Name=~s/`/\-/g;
	$Name=~s/;/\-/g;
	$Name=~s/\?/\-/g;
	
	return "$Name";
}

sub TrackFileName
{
	my $TrackNum=shift;
	my $TrackTitle=shift;
	
	$TrackTitle=FixName($TrackTitle);
	
	return sprintf("%02d - %s.ogg",$TrackNum,$TrackTitle);
}

sub ProcessSingleTrackFlac
{
	foreach $File (@_)
	{
		my $Created=0;
		
		print "Processing $File\n";
		
		($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,$size,
			$atime,$FlacMTime,$ctime,$blksize,$blocks)
				= stat($File);
		
		my $srcfile = Audio::FLAC::Header->new($File);
		my $srcframes = $srcfile->tags();

		my $MaxTrack=0;
				
		my %GlobalTags;
		
		foreach $framename (keys %$srcframes)
		{
			if ($framename ne "COVERART" && $framename ne "VENDOR")
			{
				if ($framename=~/REPLAYGAIN/)
				{
					#print "Ignoring $framename\n";
				}
				else
				{
					$GlobalTags{$framename}=$srcframes->{$framename};
				}
			}
		}

		($MP3Dir,$MP3Base)=ParseFile($File,$SourceDir);
		$MP3Dir=$DestDir."/".$MP3Dir;

		my $TrackFile="$MP3Dir/$MP3Base.ogg";
		
		if (-f "$TrackFile")
		{
			#print "File $TrackFile already exists\n";
		}
		else
		{
			$Created=1;
			print "File $TrackFile doesn't exist\n";
			
			if (!-d $MP3Dir)
			{
				MakeDirTree ($MP3Dir);
			}
			
			my $RetVal=system "flac -s -d -c \"$File\" | oggenc -Q -q 5 -o \"$TrackFile\" - ";
			#my $RetVal=system "flac -d -c \"$File\" | lame --quiet --preset fast standard - \"$TrackFile\"";
		}
		
		if (-f "$TrackFile")
		{
			($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,$size,
				$atime,$MP3MTime,$ctime,$blksize,$blocks)
					= stat($TrackFile);
			
			
			if ($MP3MTime<$FlacMTime || $Created==1)
			{
				print "Tagging $TrackFile\n";
						
				my $ogg = Audio::TagLib::Ogg::Vorbis::File->new($TrackFile);
				
				RemoveReplayGain($ogg);
				
				foreach $GlobalTag (keys %GlobalTags)
				{
					$Name = Audio::TagLib::String->new($GlobalTag,"Latin1");
					$Value = Audio::TagLib::String->new($GlobalTags{$GlobalTag},"UTF8");
					$ogg->tag()->addField($Name,$Value);
				}
		
				$ogg->save();
			}
		}
		else
		{
			print "Whoops - $TrackFile doesn't exist\n";
		}
	}
}

sub ProcessMultiTrackFlac
{
	foreach $File (@_)
	{
		my $Created=0;
		
		($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,$size,
			$atime,$FlacMTime,$ctime,$blksize,$blocks)
				= stat($File);
		
		my $srcfile = Audio::FLAC::Header->new($File);
		my $srcframes = $srcfile->tags();

		my $MaxTrack=0;
				
		my %GlobalTags;
		my @TrackTags;
		
		foreach $framename  (keys %$srcframes)
		{
			if ($framename ne "COVERART" && $framename ne "VENDOR")
			{
				if ($framename=~/REPLAYGAIN/)
				{
					#print "Ignoring $framename\n";
				}
				else
				{
					if ($framename=~/(.*)\[(.*)\]/)
					{
						if ($2>$MaxTrack)
						{
							$MaxTrack=$2;
						}
					
						$TrackTags[$2]{$1}=$srcframes->{$framename};
					}
					else
					{
						$GlobalTags{$framename}=$srcframes->{$framename};
					}
				}
			}
		}

		#foreach $tag (keys %GlobalTags)
		#{
		#	print "'" . $tag . "' is '" . $GlobalTags{$tag} . "'\n";
		#}
		
		#next;
		
		($MP3Dir,$MP3Base)=ParseFile($File,$SourceDir);
		$MP3Dir=$DestDir."/".$MP3Dir;
		
		my $TrackFile=sprintf("$MP3Dir/$MP3Base - %02d.ogg",$MaxTrack);
		
		if (-f "$TrackFile")
		{
		}
		else
		{
			$Created=1;
			
			print "File $TrackFile doesn't exist\n";
			
			if (!-d $MP3Dir)
			{
				MakeDirTree ($MP3Dir);
			}
			
			system "metaflac --export-cuesheet-to=\"$Flac.cue\" \"$Flac\"";
			my $RetVal=system "cuebreakpoints \"$Flac.cue\" 2>/dev/null | shntool split -n \"\" -d \"$MP3Dir\" -o cust ext=ogg \{ oggenc -q 5 -o %f - \}  \"$Flac\"";
			#my $RetVal=system "cuebreakpoints \"$Flac.cue\" | shntool split -n \"\" -d \"$MP3Dir\" -o cust ext=mp3 \{ lame --quiet --preset fast standard - %f \} \"$Flac\"";
			unlink "$Flac.cue";
			
			if ($RetVal==0)
			{
				for (my $Number=1;$Number<=$MaxTrack;$Number++)
				{
	        my $OldFile=sprintf("$MP3Dir/%03d.ogg",$Number);
					my $NewFile=sprintf("$MP3Dir/$MP3Base - %02d.ogg",$Number);

					print "Renaming $OldFile to $NewFile\n";
					
					unlink $NewFile;
					rename $OldFile,$NewFile;
				}
			}
		}
		
		if (-f "$TrackFile")
		{
			for (my $Track=1;$Track<=$MaxTrack;$Track++)
			{
				my $NewFile=sprintf("$MP3Dir/$MP3Base - %02d.ogg",$Track);

				($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,$size,
					$atime,$MP3MTime,$ctime,$blksize,$blocks)
						= stat($NewFile);
				
				if ($MP3MTime<$FlacMTime || $Created==1)
				{
					print "Tagging $NewFile\n";
					
					my $ogg = Audio::TagLib::Ogg::Vorbis::File->new($NewFile);

					RemoveReplayGain($ogg);					

					foreach $TrackTag (keys %{$TrackTags[$Track]})
					{
						$Name = Audio::TagLib::String->new($TrackTag,"Latin1");
						$Value = Audio::TagLib::String->new($TrackTags[$Track]{$TrackTag},"UTF8");
						$ogg->tag()->addField($Name,$Value);
					}

					foreach $GlobalTag (keys %GlobalTags)
					{
						if ($TrackTags[$Track]{$GlobalTag})
						{
							#print "Not tagging with global tag $GlobalTag - in Track Tags\n";
						}
						else
						{
							$Name = Audio::TagLib::String->new($GlobalTag,"Latin1");
							$Value = Audio::TagLib::String->new($GlobalTags{$GlobalTag},"UTF8");
							$ogg->tag()->addField($Name,$Value);
						}
					}
			
					$ogg->save();
				}
			}
		}
		else
		{
			print "Whoops, $TrackFile doesn't exist\n";
		}
	}
}

sub ScanFlacs
{
	my @Dirs;
	
	foreach my $SourceDir (@_)
	{
		opendir FLACDIR,$SourceDir;
		while (my $DirEntry=readdir(FLACDIR))
		{
			if (($DirEntry ne ".") && ($DirEntry ne "..") && -d "$SourceDir/$DirEntry")
			{
				$Dirs[$#Dirs + 1]="$SourceDir/$DirEntry";
			}
			
			$Flacs[$#Flacs + 1]="$SourceDir/$DirEntry" if $DirEntry=~/\.flac$/;
		}
	}
	
	foreach my $Dir (@Dirs)
	{
		ScanFlacs($Dir);
	}
}

sub SingleTrackFlac
{
	my $Flac=shift;
	
	$RetVal=system("metaflac --export-cuesheet-to=- \"$Flac\" 2>/dev/null >/dev/null");
	if ($RetVal==0)
	{
		return 0;
	}
	else
	{
		return 1;
	}
}

sub ProcessFlac
{
	foreach my $Flac (@_)
	{
		if (SingleTrackFlac($Flac))
		{
			ProcessSingleTrackFlac($Flac);
		}
		else
		{
			ProcessMultiTrackFlac($Flac);
		}
	}
}

@Flacs=();

$SourceDir=$ARGV[0];
$DestDir=$ARGV[1];

ScanFlacs($SourceDir);

$Count=1;

@Sorted=sort(@Flacs);

foreach $Flac (@Sorted)
{
	print $Count;
	print " of ";
	print $#Flacs+1;
	print " - ";
	print $Flac;
	print "\n";
	$Count++;
	
	ProcessFlac($Flac);
}

File: flac2ogg.py
#!/usr/bin/python

import sys
import os
import mutagen.flac
import mutagen.oggvorbis
import re
import shutil
import filecmp

def ExtractCoverArt(FlacFile,MP3Dir):
	CoverArt=os.path.join(os.path.dirname(FlacFile),"cover.jpg")
	print "Cover art is " + CoverArt

	DestFile=os.path.join(MP3Dir,"Cover.jpg")
	
	if os.path.exists(CoverArt):
		if os.path.ctime(CoverArt) > os.path.ctime(DestFile):
			print "Copying cover art file"
			shutil.copyfile(CoverArt,DestFile)
		else:
			print "Cover art hasn't changed"
	else:
		TmpFile=os.path.join(os.path.dirname(FlacFile),"tmpcover.jpg")
		
		print "Extracting cover art from FLAC to " + TmpFile
		os.system("metaflac --export-picture-to=\"" + TmpFile + "\" \"" + FlacFile + "\"")
		
		if os.path.exists(TmpFile):
			if not os.path.exists(DestFile) or not filecmp.cmp(TmpFile,DestFile):
				print "Copying extracted cover art file"
				shutil.copyfile(TmpFile,DestFile)
			else:
				print "Extracted cover art matches the one already there"
				
			os.unlink(TmpFile)					
		else:
			print TmpFile + " doesn't exist after extraction"

def ProcessSingleTrackFlac(basedir,file,destdir):
	FullPath=os.path.join(basedir,file)
		
	MaxTrack=0
	GlobalTags={}
			
	FlacMetadata=mutagen.flac.Open(FullPath)

	for key in FlacMetadata.keys():
		if key!= "coverart" and key != "vendor":
			if not key.startswith("replaygain"):
				GlobalTags[key]=FlacMetadata[key];

	MP3Dir=os.path.join(destdir,os.path.dirname(file))
	(dummy,MP3Base)=os.path.split(file)
	(MP3Base,dummy)=os.path.splitext(MP3Base)

	TrackFile=os.path.join(MP3Dir,MP3Base)+".ogg"

	if not os.path.exists(TrackFile):
		if not os.path.exists(os.path.dirname(TrackFile)):
			os.makedirs(os.path.dirname(TrackFile))

		print "Extracting " + FullPath
		os.system("flac -s -d -c \"" + FullPath + "\" | oggenc -Q -q 5 -o \"" + TrackFile + "\" - ");

	if os.path.exists(TrackFile):
		ActualFileTags = LoadOggFileTags(TrackFile)
		
		if GlobalTags != ActualFileTags:
			print "Tagging " + TrackFile
			WriteOggTags(TrackFile,GlobalTags)
	else:
		print "Whoops - " + TrackFile + "doesn't exist"

	ExtractCoverArt(FullPath,MP3Dir)

def ProcessMultiTrackFlac(basedir,file,destdir):
	FullPath=os.path.join(basedir,file)
		
	MaxTrack=0
	GlobalTags={}
	TrackTags={}
	
	FlacMetadata=mutagen.flac.Open(FullPath)
		
	for key in FlacMetadata.keys():
		if key!="coverart" and key!="vendor":
			if not key.startswith("replaygain"):
				match=re.search("(.*)\[(.*)\]",key)
				if match!=None:
					if int(match.group(2))>MaxTrack:
						MaxTrack=int(match.group(2));

					if TrackTags.get(int(match.group(2)),-1)==-1:
						TrackTags[int(match.group(2))]=dict()
									
					TrackTags[int(match.group(2))][match.group(1)]=FlacMetadata[key];
				else:
					GlobalTags[key]=FlacMetadata[key]

	MP3Dir=os.path.join(destdir,os.path.dirname(file))
	(dummy,MP3Base)=os.path.split(file)
	(MP3Base,dummy)=os.path.splitext(MP3Base)

	TrackFile=os.path.join(MP3Dir,MP3Base)
	TrackFile+=" - %02d" %MaxTrack + ".ogg"

	if not os.path.exists(TrackFile):
		if not os.path.exists(os.path.dirname(TrackFile)):
			os.makedirs(os.path.dirname(TrackFile))
			
		print "Extracting " + FullPath
		os.system("metaflac --export-cuesheet-to=\"" + FullPath + ".cue\" \"" + FullPath + "\"")
		RetVal=os.system("cuebreakpoints \"" + FullPath + ".cue\" 2>/dev/null | shntool split -O always -t \"\" -n \"%02d\" -d \"" + MP3Dir + "\" -o 'cust ext=ogg oggenc -q 5 -o %f - '  \"" + FullPath + "\"")
		os.unlink(FullPath+".cue")
		
		if RetVal==0:
			for Number in range(MaxTrack):
				OldFile=os.path.join(MP3Dir,"%02d" %(Number+1)) + ".ogg"
				NewFile=os.path.join(MP3Dir,MP3Base) + " - %02d" %(Number+1) + ".ogg";
				
				if os.path.exists(NewFile):
					os.unlink(NewFile)
					
				os.rename(OldFile,NewFile)
				
	GlobalTagKeys=GlobalTags.keys()

	for Number in range(MaxTrack):
		TrackFile=os.path.join(MP3Dir,MP3Base)
		TrackFile+=" - %02d" %(Number+1)+ ".ogg"
		
		RequiredFileTags=dict()
		
		TrackTagKeys=TrackTags[Number+1].keys()
		
		for Key in TrackTagKeys:
			RequiredFileTags[Key]=TrackTags[Number+1][Key]

		for Key in GlobalTagKeys:
			if Key not in TrackTags[Number+1]:
				RequiredFileTags[Key]=GlobalTags[Key]
			
		ActualFileTags = LoadOggFileTags(TrackFile)
		
		if RequiredFileTags != ActualFileTags:
			print "Tagging " + TrackFile
			WriteOggTags(TrackFile,RequiredFileTags)

	ExtractCoverArt(FullPath,MP3Dir)

def WriteOggTags(file,tags):
	metadata = mutagen.oggvorbis.Open(file)
	metadata.delete()
	
	keys = tags.keys()
	for key in keys:
		metadata[key]=tags[key]
		
	metadata.save()
	
def LoadOggFileTags(file):
	ret = dict()
	metadata = mutagen.oggvorbis.Open(file)
	for key in metadata.keys():
		if not key.startswith("replaygain"):
			ret[key]=metadata[key]
	
	return ret

def SingleTrackFlac(file):
	ret=os.system("metaflac --export-cuesheet-to=- \"" + file + "\" 2>/dev/null >/dev/null")
	if (ret==0):
		return False
	else:
		return True

def ProcessFlac(basedir,file,destdir):
	fullpath=os.path.join(basedir,file)
	if SingleTrackFlac(fullpath):
		ProcessSingleTrackFlac(basedir,file,destdir)
	else:
		ProcessMultiTrackFlac(basedir,file,destdir)
		
def GetFlacFiles(basedir, extradir="", files=None):
	if files==None:
		retfiles=[]
	else:
		retfiles=files
		
	thisdir=os.path.join(basedir,extradir)
	
	entries=os.listdir(thisdir)
	
	for entry in entries:
		if os.path.isdir(os.path.join(thisdir,entry)):
			GetFlacFiles(basedir,os.path.join(extradir,entry),retfiles)
		else:
			(root,ext)=os.path.splitext(entry)
			
			if (ext==".flac"):
				retfiles.append(os.path.join(extradir,entry))
		
	return retfiles
	
if len(sys.argv)==3:
	srcdir=sys.argv[1]
	destdir=sys.argv[2]
	
	files=GetFlacFiles(srcdir)

	Count=1
	for file in files:
		print str(Count) + " of " + str(len(files)) + " - " + file
		Count=Count+1
		
		ProcessFlac(srcdir,file,destdir)
else:
	print "Usage: " + sys.argv[0] + " srcdir destdir"
